### PR TITLE
fix: App delete from ACD with non cascade

### LINKED ISSUE #3304: Bug: Deleting an app for the cluster which is no longer accessible gets stuck in deleting
### 📜 Description

Deleting an app for the cluster which is no longer accessible gets stuck in deleting

### 👟 Reproduction steps

1. Add a cluster on devtron
2. Add some pipelines using gitops and do build and deploy on an environment created on that cluster
3. Delete the cluster from cloud so that the endpoint is no longer accessible from devtron or argocd
4. Try deleting the cd pipelines or helm apps from devtron for the environments on that particular cluster

### 👍 Expected behavior

Pipelines should get deleted and apps should be deleted non-cascade from argocd

### 👎 Actual Behavior

Pipelines/Apps gets stuck in deleting because those aren't able to delete from argocd

### ☸ Kubernetes version

EKS 1.24

### Cloud provider

<details>
AWS
</details>

### 🌍 Browser

Chrome

### 🧱 Your Environment

_No response_

### ✅ Proposed Solution

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find any similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#2821

### PR DESCRIPTION
**Pull request by Ash-exp**
# Description

Deleting an app for the cluster which is no longer accessible gets stuck in deleting for Argocd deployments. It is because Argocd doesn't get cluster resources for cascade delete. In this case we are deleting pipeline/ deployment(charts) in noncascade delete method from Argocd.

Fixes #3304
Ticket [AB 2821](https://dev.azure.com/DevtronLabs/Devtron/_boards/board/t/OSS%20adoption/Stories/?workitem=2821)
TechDoc [Link](https://docs.google.com/document/d/1RNin5ob-5mayKX4jfwGu9sbuDoETddBtAjDBya-13y4/edit?usp=sharing)
Frontend PR [Link](https://github.com/devtron-labs/dashboard/pull/989)

## How Has This Been Tested?
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-d89802db-7fff-da9e-2e19-6a930b656602"><div dir="ltr" style="margin-left:0pt;" align="center">

App Type | Test Case | App Deletion Status | Non-case Delete | CD Pipeline Patch (With Cascade: false) | Force Delete
-- | -- | -- | -- | -- | --
Helm App (ArgoCD Deployed) | Cluster Reachable | PARTIAL DELETE State | Disabled | Disabled | Disabled
Other State | Disabled | Disabled | Disabled
Cluster Unreachable | PARTIAL DELETE State | Enabled | Disabled | Disabled
Other State | Disabled | Enabled | Disabled
Devtron App (ArgoCD Deployed) | Cluster Reachable | PARTIAL DELETE State | Disabled | Disabled | Disabled
Other State | Disabled | Disabled | Disabled
Cluster Unreachable | PARTIAL DELETE State | Enabled | Disabled | Disabled
Other State | Disabled | Enabled | Disabled
Helm App (Helm Deployed) | ANY | ANY | Disabled | Disabled | Disabled
Devtron App (Helm Deployed) | ANY | ANY | Disabled | Disabled | Disabled
ANY | Deletion Error (Other than Cluster Unreachable) | ANY | Disabled | Disabled | Enabled

</div></b>

- [x] For Devtron App
  - [x] Cluster has just stopped (within 15 mins)
  - [x] Cluster has been stopped (more than 15 mins)
  - [x] service manifest Finaliser manipulation (Should not delete App)
- [x] For Helm App
  - [x] Cluster has just stopped (within 15 mins)
  - [x] Cluster has been stopped (more than 15 mins)
  - [x] service manifest Finaliser manipulation (Should not delete App)

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: util/k8s/k8sApplicationService.go [hunk 1]
```diff
@@ -811,3 +814,14 @@ func (impl *K8sApplicationServiceImpl) applyResourceFromManifest(ctx context.Con
 
 	return isUpdateResource, nil
 }
+
+func (impl *K8sApplicationServiceImpl) FetchConnectionStatusForCluster(k8sClientSet *kubernetes.Clientset, clusterId int) error {
+	//using livez path as healthz path is deprecated
+	path := "/livez"
+	response, err := k8sClientSet.Discovery().RESTClient().Get().AbsPath(path).DoRaw(context.Background())
+	log.Println("received response for cluster livez status", "response", string(response), "err", err, "clusterId", clusterId)
```
Comment (prakash100198) [inline util/k8s/k8sApplicationService.go:822]: Remove this Println message, and also handle error here for err!=nil

File: pkg/bulkAction/BulkUpdateService.go [hunk 2]
```diff
@@ -1515,7 +1515,13 @@ func (impl BulkUpdateServiceImpl) PerformBulkActionOnCdPipelines(dto *CdBulkActi
 	ctx context.Context, dryRun bool, impactedAppWfIds []int, impactedCiPipelineIds []int) (*PipelineAndWfBulkActionResponseDto, error) {
 	switch dto.Action {
 	case CD_BULK_DELETE:
-		bulkDeleteResp, err := impl.PerformBulkDeleteActionOnCdPipelines(impactedPipelines, ctx, dryRun, dto.ForceDelete, dto.DeleteWfAndCiPipeline, impactedAppWfIds, impactedCiPipelineIds, dto.UserId)
+		deleteAction := bean2.CASCADE_DELETE
+		if dto.ForceDelete {
+			deleteAction = bean2.FORCE_DELETE
+		} else if dto.CascadeDelete {
+			deleteAction = bean2.NON_CASCADE_DELETE
```
Comment (prakash100198) [inline pkg/bulkAction/BulkUpdateService.go:1522]: if dto has cascade delete action then why delete action is set to NON_CASCADE_DELETE?

**Review by prakash100198 — CHANGES_REQUESTED** [review]

File: api/appStore/InstalledAppRestHandler.go [hunk 3]
```diff
@@ -429,6 +434,82 @@ func (handler *InstalledAppRestHandlerImpl) FetchNotesForArgoInstalledApp(w http
...
+		return
+	}
+	if util.IsBaseStack() || util.IsHelmApp(installedApp.AppOfferingMode) || util2.IsHelmApp(installedApp.DeploymentAppType) {
+		handler.Logger.Errorw("request err, NonCascadeDeleteCdPipeline", "err", fmt.Errorf("nocascade delete is not supported for %s", installedApp.AppName))
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+	//rbac block starts from here
+	rbacObject, rbacObject2 := handler.enforcerUtil.GetHelmObjectByAppNameAndEnvId(installedApp.AppName, installedApp.EnvironmentId)
+	ok := handler.enforcer.Enforce(token, casbin.ResourceHelmApp, casbin.ActionDelete, rbacObject) || handler.enforcer.Enforce(token, casbin.ResourceHelmApp, casbin.ActionDelete, rbacObject2)
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply api/appStore/InstalledAppRestHandler.go:477]: yeah tested across different users

File: api/restHandler/app/PipelineConfigRestHandler.go [hunk 4]
```diff
@@ -234,6 +236,88 @@ func (handler PipelineConfigRestHandlerImpl) DeleteApp(w http.ResponseWriter, r
...
+	if err != nil && err != pg.ErrNoRows {
+		handler.Logger.Errorw("error in fetching pipelines from db", "appId", appId, "envId", envId)
+		common.WriteJsonResp(w, err, "error in fetching pipelines from db", http.StatusInternalServerError)
+		return
+	}
+	if len(pipelines) == 0 {
+		common.WriteJsonResp(w, err, "deployment not found, unable to fetch resource tree", http.StatusNotFound)
+		return
+	}
+	if len(pipelines) > 1 {
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply api/restHandler/app/PipelineConfigRestHandler.go]: Implemented 👍

File: pkg/appStore/deployment/service/AppStoreDeploymentService.go [hunk 5]
```diff
@@ -561,31 +561,18 @@ func (impl AppStoreDeploymentServiceImpl) GetAllInstalledAppsByAppStoreId(w http
 	}
 	var installedAppsEnvResponse []appStoreBean.InstalledAppsResponse
 	for _, a := range installedApps {
-		var status string
-		if util2.IsBaseStack() || util2.IsHelmApp(a.AppOfferingMode) || util.IsHelmApp(a.DeploymentAppType) {
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply pkg/appStore/deployment/service/AppStoreDeploymentService.go:657]: here we were fetching live status, I replaced it with the DB app status values.

File: pkg/appStore/deployment/tool/gitops/AppStoreDeploymentArgoCdService.go [hunk 6]
```diff
@@ -675,14 +677,15 @@ func (impl AppStoreDeploymentArgoCdServiceImpl) updateValuesYaml(environment *cl
...
+	var err error
+	err = impl.deleteACD(acdAppName, ctx, installAppVersionRequest.NonCascadeDelete)
 	if err != nil {
 		impl.Logger.Errorw("error in deleting ACD ", "name", acdAppName, "err", err)
 		if installAppVersionRequest.ForceDelete {
 			impl.Logger.Warnw("error while deletion of app in acd, continue to delete in db as this operation is force delete", "error", err)
 		} else {
 			//statusError, _ := err.(*errors2.StatusError)
-			if strings.Contains(err.Error(), "code = NotFound") {
+			if !installAppVersionRequest.NonCascadeDelete && strings.Contains(err.Error(), "code = NotFound") {
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply pkg/appStore/deployment/tool/gitops/AppStoreDeploymentArgoCdService.go:634]: Its should be this way to handle cascade flow error

File: util/k8s/k8sApplicationService.go (hunk 1 above)
Comment (Ash-exp) [reply util/k8s/k8sApplicationService.go:869]: it was called through a go routine and saving the error in db

File: pkg/bulkAction/BulkUpdateService.go (hunk 2 above)
Comment (Ash-exp) [reply pkg/bulkAction/BulkUpdateService.go:1522]: updated 👍

File: api/cluster/EnvironmentRestHandler.go [hunk 7]
```diff
@@ -460,3 +472,73 @@ func (impl EnvironmentRestHandlerImpl) GetCombinedEnvironmentListForDropDownByCl
...
+	}
+	err = impl.k8sApplicationService.FetchConnectionStatusForCluster(k8sClientSet, clusterBean.Id)
+	if err != nil {
+		responseObj.ClusterReachable = false
+	}
+	//updating the cluster connection error to db
+	mapObj := map[int]error{
+		clusterBean.Id: err,
+	}
+	_ = impl.environmentClusterMappingsService.HandleErrorInClusterConnections(mapObj)
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply api/cluster/EnvironmentRestHandler.go]: doesn't require error catching as it's independent of the service.(updating db for cluster conn error)

**Review by iamayushm — APPROVED** [review]

**Review by prakash100198 — APPROVED** [review]

**Pull request merged**